<a href="https://colab.research.google.com/github/danavanwyk/Machine-Learning/blob/main/Intro%20to%20Spacy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Importing Spacy and assigning it a variable name to ensure ease of use throughout our process. 


In [3]:
import spacy

nlp = spacy.load('en_core_web_sm')

We will be applying, Tokenisation, Lemmatisation and then entity recognition Using the spacy english model. We have imported some "Garden path sentences" fro: https://www.apartmenttherapy.com/garden-sentences-262915 and assigned each a recognizeable variable name. 

In [31]:
old_man_boat = "The old man the boat."
soldiers_house = "The complex houses married and single soldiers and their families."
horse_barn = "The horse raced past, the barn fell."
florist = "The florist sent the flowers was pleased."
cotton_clothing = "The cotton clothing is made of grows in Mississippi."
sour_ocean = "The sour drink from the ocean."

All of our sentences and variable names: 

In [32]:
print(f"old_man_boat = {old_man_boat}\nsoldiers_house = {soldiers_house}\nhorse_barn = {horse_barn}\nflorist = {florist}\ncotton_clothing = {cotton_clothing}\nsour_ocean = {sour_ocean} ")

old_man_boat = The old man the boat.
soldiers_house = The complex houses married and single soldiers and their families.
horse_barn = The horse raced past, the barn fell.
florist = The florist sent the flowers was pleased.
cotton_clothing = The cotton clothing is made of grows in Mississippi.
sour_ocean = The sour drink from the ocean. 


Now that we have our sentences stored inside variables we can start with the process of Tokenisations: 

Tokenisation is a foundational step in many NLP tasks. Tokenising text is the
process of splitting a piece of text into words, symbols, punctuation, spaces
and other elements, thereby creating “tokens”. A naive way to do this is to
simply split the string on white space:

In [33]:
#Applying Natural Language processing to each of our sentences. We stores the spacy model in the variable name nlp earlier. 

tok_old_man_boat = nlp(old_man_boat)
tok_soldiers_house = nlp(soldiers_house)
tok_horse_barn = nlp(horse_barn)
tok_florist = nlp(florist)
tok_cotton_clothing = nlp(cotton_clothing)
tok_sour_ocean = nlp(sour_ocean)


We will be using the split function to get each of our sentences,which are currently of data type string into lists. 

In [34]:
#"the old man the boat."
tok_old_man_boat.text.split()


['The', 'old', 'man', 'the', 'boat.']

At first glance this look fine, but if you look closer it fails to recognise the full stop and other elements of the sentence that helps us understand it structure. 

In [24]:
[token.orth_ for token in tok_old_man_boat]

['The', 'old', 'man', 'the', 'boat', '.']

Here we access the each token’s .orth_ method, which returns a string representationof the token rather than a SpaCy token object, this might not always be desirable,but worth noting. SpaCy recognises punctuation and is able to split these punctuation tokens from word tokens

In [35]:
#"The complex houses married and single soldiers and their families."
tok_soldiers_house.text.split() 

['The',
 'complex',
 'houses',
 'married',
 'and',
 'single',
 'soldiers',
 'and',
 'their',
 'families.']

In [36]:
[token.orth_ for token in tok_soldiers_house]

['The',
 'complex',
 'houses',
 'married',
 'and',
 'single',
 'soldiers',
 'and',
 'their',
 'families',
 '.']

In [37]:
#"The horse raced past the barn fell."
tok_horse_barn.text.split()

['The', 'horse', 'raced', 'past,', 'the', 'barn', 'fell.']

In [38]:
{token.orth_ for token in tok_horse_barn}

{',', '.', 'The', 'barn', 'fell', 'horse', 'past', 'raced', 'the'}

In [39]:
#"The florist sent the flowers was pleased."
tok_florist.text.split()

['The', 'florist', 'sent', 'the', 'flowers', 'was', 'pleased.']

In [40]:
[token.orth_ for token in tok_florist]

['The', 'florist', 'sent', 'the', 'flowers', 'was', 'pleased', '.']

In [41]:
#"The cotton clothing is made of grows in Mississippi.""
tok_cotton_clothing.text.split()

['The',
 'cotton',
 'clothing',
 'is',
 'made',
 'of',
 'grows',
 'in',
 'Mississippi.']

In [44]:
[token.orth_ for token in tok_cotton_clothing]

['The',
 'cotton',
 'clothing',
 'is',
 'made',
 'of',
 'grows',
 'in',
 'Mississippi',
 '.']

In [45]:
tok_sour_ocean.text.split()

['The', 'sour', 'drink', 'from', 'the', 'ocean.']

In [46]:
[token.orth_ for token in tok_sour_ocean]

['The', 'sour', 'drink', 'from', 'the', 'ocean', '.']

Many of SpaCy’s token method offer both string and integer representations of processed text – methods with an underscore suffix return strings, methods without an underscore suffix return integers. For example:

In [48]:
print([(token, token.orth_, token.orth) for token in tok_old_man_boat])
print([(token, token.orth_, token.orth) for token in tok_horse_barn])
print([(token, token.orth_, token.orth) for token in tok_florist])
print([(token, token.orth_, token.orth) for token in tok_cotton_clothing])
print([(token, token.orth_, token.orth) for token in tok_soldiers_house])
print([(token, token.orth_, token.orth) for token in tok_sour_ocean])


[(The, 'The', 5059648917813135842), (old, 'old', 2483095116303079762), (man, 'man', 3104811030673030468), (the, 'the', 7425985699627899538), (boat, 'boat', 11387929873876272110), (., '.', 12646065887601541794)]
[(The, 'The', 5059648917813135842), (horse, 'horse', 9866490340549498003), (raced, 'raced', 11985977600559752646), (past, 'past', 14080206577736366893), (,, ',', 2593208677638477497), (the, 'the', 7425985699627899538), (barn, 'barn', 1731024304121373681), (fell, 'fell', 17406646348554510309), (., '.', 12646065887601541794)]
[(The, 'The', 5059648917813135842), (florist, 'florist', 10501838060908119271), (sent, 'sent', 16056986141218963985), (the, 'the', 7425985699627899538), (flowers, 'flowers', 2477902471846530091), (was, 'was', 9921686513378912864), (pleased, 'pleased', 4426875666208874359), (., '.', 12646065887601541794)]
[(The, 'The', 5059648917813135842), (cotton, 'cotton', 13158503143813622304), (clothing, 'clothing', 11978358972660538478), (is, 'is', 3411606890003347522), 

OUTPUT:
[(The, 'The', 5059648917813135842), (old, 'old', 2483095116303079762), (man, 'man', 3104811030673030468), (the, 'the', 7425985699627899538), (boat, 'boat', 11387929873876272110), (., '.', 12646065887601541794)]


[(The, 'The', 5059648917813135842), (horse, 'horse', 9866490340549498003), (raced, 'raced', 11985977600559752646), (past, 'past', 14080206577736366893), (,, ',', 2593208677638477497), (the, 'the', 7425985699627899538), (barn, 'barn', 1731024304121373681), (fell, 'fell', 17406646348554510309), (., '.', 12646065887601541794)]


[(The, 'The', 5059648917813135842), (florist, 'florist', 10501838060908119271), (sent, 'sent', 16056986141218963985), (the, 'the', 7425985699627899538), (flowers, 'flowers', 2477902471846530091), (was, 'was', 9921686513378912864), (pleased, 'pleased', 4426875666208874359), (., '.', 12646065887601541794)]


[(The, 'The', 5059648917813135842), (cotton, 'cotton', 13158503143813622304), (clothing, 'clothing', 11978358972660538478), (is, 'is', 3411606890003347522), (made, 'made', 7136294487848412980), (of, 'of', 886050111519832510), (grows, 'grows', 6115445568100627063), (in, 'in', 3002984154512732771), (Mississippi, 'Mississippi', 1946980641541756374), (., '.', 12646065887601541794)]


[(The, 'The', 5059648917813135842), (complex, 'complex', 11125307656307208867), (houses, 'houses', 12236405033243632862), (married, 'married', 15974428657194256046), (and, 'and', 2283656566040971221), (single, 'single', 9644341664820045211), (soldiers, 'soldiers', 13157451586059593190), (and, 'and', 2283656566040971221), (their, 'their', 4244585616942201722), (families, 'families', 13150649422924114104), (., '.', 12646065887601541794)]


[(The, 'The', 5059648917813135842), (sour, 'sour', 1437310558085749269), (drink, 'drink', 8799542030263951485), (from, 'from', 7831658034963690409), (the, 'the', 7425985699627899538), (ocean, 'ocean', 11641983324036055121), (., '.', 12646065887601541794)]

We can also exclude punctuatiuon and white space with the following methods included in spacy. is_punct and is_space

In [58]:
print([token.orth_ for token in tok_old_man_boat if not token.is_punct | token.is_space ])
print([token.orth_ for token in tok_horse_barn if not token.is_punct | token.is_space])
print([token.orth_ for token in tok_soldiers_house if not token.is_punct | token.is_space])
print([token.orth_ for token in tok_cotton_clothing if not token.is_punct |token.is_space])
print([token.orth_ for token in tok_florist if not token.is_punct | token.is_space])
print([token.orth_ for token in tok_sour_ocean if not token.is_punct | token.is_space])

['The', 'old', 'man', 'the', 'boat']
['The', 'horse', 'raced', 'past', 'the', 'barn', 'fell']
['The', 'complex', 'houses', 'married', 'and', 'single', 'soldiers', 'and', 'their', 'families']
['The', 'cotton', 'clothing', 'is', 'made', 'of', 'grows', 'in', 'Mississippi']
['The', 'florist', 'sent', 'the', 'flowers', 'was', 'pleased']
['The', 'sour', 'drink', 'from', 'the', 'ocean']


We can identify stop word with the is.stop method. Let's use a for loop to iterate through the list. 

We can create a simple function to call each of our garden path sentences. 

In [71]:
def stop_method(sentence):
  for each_word in sentence:
    if each_word.is_stop ==True:
      print(each_word)


In [73]:
stop_method(tok_old_man_boat)

The
the
The
from
the


In [74]:
stop_method(tok_soldiers_house)

The
and
and
their


In [75]:
stop_method(tok_horse_barn)

The
the


In [77]:
stop_method(tok_sour_ocean)

The
from
the


In [78]:
stop_method(tok_florist)

The
the
was


Now let's get on to Lemmatisation.  Lemmatisation is the process of reducing a word to its base form. We will be creating a simple function to be able to reus the code. 

Some of these might look a little strange but let's comare it to the original sentence and see what spacy is actually doing. 


In [95]:
def practical_lemma(sentence):
  print(sentence)
  print([word.lemma_ for word in sentence if not word.is_punct | word.is_space])

In [96]:
practical_lemma(tok_old_man_boat)

The old man the boat.
['the', 'old', 'man', 'the', 'boat']


In [97]:
practical_lemma(tok_horse_barn)

The horse raced past, the barn fell.
['the', 'horse', 'race', 'past', 'the', 'barn', 'fall']


In [98]:
practical_lemma(tok_soldiers_house)

The complex houses married and single soldiers and their families.
['the', 'complex', 'house', 'marry', 'and', 'single', 'soldier', 'and', 'their', 'family']


In [99]:
practical_lemma(tok_sour_ocean)

The sour drink from the ocean.
['the', 'sour', 'drink', 'from', 'the', 'ocean']


In [100]:
practical_lemma(tok_florist)

The florist sent the flowers was pleased.
['the', 'florist', 'send', 'the', 'flower', 'be', 'pleased']


In [101]:
practical_lemma(tok_cotton_clothing)

The cotton clothing is made of grows in Mississippi.
['the', 'cotton', 'clothing', 'be', 'make', 'of', 'grow', 'in', 'Mississippi']


As we can see it transforms the word to it's base form. You might wonder why this is usefull..

An immediate use case is in machine learning, specifically text classification. Lemmatising the text prior to, for example, creating a“bag-of-words” avoids word duplication and, therefore, allows for the model to build a clearer picture of patterns of word usage across multiple documents.

Entity Recognition: This is where it start to get rea\:ly freaky!!

Spacy shows off it capabilities by classifying named entities found in a text into pre-defined categories, such as persons, places, organisations, dates,etc. spaCy uses a statistical model to classify a broad range of entities,including persons, events, works-of-art and nationalities / religions (see the documentation for the full list https://spacy.io/docs/usage/entity-recognition).

In [103]:
#Same thing let's first create a simple fucntion so we are not typing out everything over and over. Although it might be good for practice to type it out..
def ent_recog_gps(sentence):
  print(sentence)
  print([[(i, i.label_, i.label) for i in sentence.ents]])

In [105]:
ent_recog_gps(tok_old_man_boat)

The old man the boat.
[[]]


In [106]:
ent_recog_gps(tok_horse_barn)

The horse raced past, the barn fell.
[[]]


In [108]:
ent_recog_gps(tok_cotton_clothing)

The cotton clothing is made of grows in Mississippi.
[[(Mississippi, 'GPE', 384)]]


In [109]:
ent_recog_gps(tok_sour_ocean)

The sour drink from the ocean.
[[]]


In [110]:
ent_recog_gps(tok_florist)

The florist sent the flowers was pleased.
[[]]


In [111]:
ent_recog_gps(tok_soldiers_house)

The complex houses married and single soldiers and their families.
[[]]


As we can see here not much to recognise in these short sentences, But let's apply it to a larger sentence containing a few more recognisable entities. 

In [114]:
deadpool = "Wade Wilson (Ryan Reynolds) is a former Special Forces operative who now works as a mercenary. His world comes crashing down when evil scientist Ajax (Ed Skrein) tortures, disfigures and transforms him into Deadpool. The rogue experiment leaves Deadpool with accelerated healing powers and a twisted sense of humor."
#tokenization
tok_deadpool = nlp(deadpool)
tok_deadpool.text.split()

ent_recog_gps(tok_deadpool)

Wade Wilson (Ryan Reynolds) is a former Special Forces operative who now works as a mercenary. His world comes crashing down when evil scientist Ajax (Ed Skrein) tortures, disfigures and transforms him into Deadpool. The rogue experiment leaves Deadpool with accelerated healing powers and a twisted sense of humor.
[[(Wade Wilson, 'PERSON', 380), (Ryan Reynolds, 'PERSON', 380), (Ajax, 'PERSON', 380), (Ed Skrein, 'PERSON', 380), (Deadpool, 'ORG', 383), (Deadpool, 'ORG', 383)]]


Much better:

Wade Wilson (Ryan Reynolds) is a former Special Forces operative who now works as a mercenary. His world comes crashing down when evil scientist Ajax (Ed Skrein) tortures, disfigures and transforms him into Deadpool. The rogue experiment leaves Deadpool with accelerated healing powers and a twisted sense of humor.


[[(Wade Wilson, 'PERSON', 380), (Ryan Reynolds, 'PERSON', 380), (Ajax, 'PERSON', 380), (Ed Skrein, 'PERSON', 380), (Deadpool, 'ORG', 383), (Deadpool, 'ORG', 383)]]